### ISSUE TITLE
Reverse routing params conforming

### ISSUE THREAD

**Issue by malchmih** [metosin/reitit#75]
I'd be useful to have coercion to work with `reitit/match-by-name`. One example is conforming keyword to a string path parameter.
Related `spec-tools` issue: https://github.com/metosin/spec-tools/issues/112

```clj
(r/match-by-name
  (r/router
    ["/olipa/:kerran" ::avaruus])
  ::avaruus
  {:kerran :joskus})
;#Match{:template "/olipa/:kerran",
;       :data {:name :user/avaruus},
;       :result nil,
;       :path-params {:kerran :joskus},   <-- wrong
;       :path "/olipa/:joskus"}           <-- wrong
```

**Comment by ikitommi** [comment]
This should be polymorphic. Alternatives:

# 1) Use JSON encoder

* most likely does the job already.
* requires dependency to JSON library
* compojure-api does this

# 2) Multimethod

* e.g. `reitit.core/stringify`
* Easy to implement
* Value-based, works for all `coercion` instances
* Slow(er than Protocol)
* Adding new types (e.g. `LocalTime`) requires imperative registration.
  * Extending just for routing, can't (shoudn't) use it elsewhere

# 3) Protocol

* e.g. `reitit.core/Stringify`
* Easy to implement
* Value-based, works for all `coercion` instances
* Fast(er than Multimethod)
* Adding new types (e.g. `LocalTime`) is more declarative
  * Extending just for routing, can't (shoudn't) use it elsewhere
* bidi does this

# 4) Coercion/Type-based

* bijections, defined for all types of coercion
* both coercion & uncoercion defined in the type meta data
* works for all use cases, not just routing (where else this is really needed?)
* parameter type needs to be defined in order to work
* e.g. with spec tools to `keyword?`:

```clj
(st/spec
  {:spec keyword?
   :description "a bijecting keyword"
   ::unform/string #(name %2)
   ::conform/string #(keyword %2)})
```

**Comment by malchmih** [comment]
From my point of view, 3rd option (Protocol) is the more natural choice here, since I don't see any other usage of this except for the routing

**Comment by ikitommi** [comment]
I think it's the way to go. When/if 4rd is implemented, it can be used instead. Would you like to do a PR?

**Comment by ikitommi** [comment]
I have extra time just now, will do this.

**Comment by ikitommi** [comment]
#76

**Comment by ikitommi** [comment]
fixed in `[metosin/reitit-core "0.1.1-20180322.170157-5"]`.
